core/fortran-example/tutorial/02_tuple/main.cpp

```diff
@@ -0,0 +1,113 @@
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include <Cabana_Core.hpp>
+
+#include <iostream>
+
+//---------------------------------------------------------------------------//
+// Tuple example.
+//---------------------------------------------------------------------------//
+
+/*
+      Cabana tuples are similar to C++11 tuples in that they are ordered lists
+      of elements of different types specified at compile time. Some aspects
+      of Cabana::Tuple that are different from std::tuple:
+
+      *: Cabana tuples must consist of data of trivial types and classes. A
+         trivial type or class is one who's
+
+         1) storage is contiguous (i.e. trivially copyable)
+
+         2) which only supports static default initialization (trivially
+            default constructible), either cv-qualified or not.
+
+         Trivial types include scalar types, trivial classes and arrays of any
+         such types. A trivial class is a class (defined with class, struct or
+         union) that is both trivially default constructible and trivially
+         copyable, which implies that:
+
+         1) uses the implicitly defined default copy and move constructors,
+            copy and move assignments, and destructor
+
+         2) has no virtual members
+
+         3) has no non-static data members with brace- or equal- initializers
+
+         4) its base class and non-static data members (if any) are themselves
+            also trivial types
+
+      *: The types in a Cabana tuple are defined via a type parameter
+         pack using the Cabana::MemberTypes class
+
+      *: Cabana tuples are designed to store and access multidimensional array
+         data with accessors provided to obtain references to individual array
+         values within the tuple.
+*/
+
+/* Start by declaring the types in our tuple will store. Store a rank-2
+       array of doubles, a rank-1 array of floats, and a single integer in
+       each tuple.
+
+*/
+using DataTypes = Cabana::MemberTypes<double[3][3],
+                                          float[4],
+                                          int>;
+
+/*
+    Create the tuple. This tuple is identical to:
+
+     struct local_particle_struct_t
+     {
+           double d0[3][3];
+           float d1[4];
+           int d2;
+     };
+
+
+       Note that ata members in Cabana tuples are stored in the same order in
+       which they are declared.
+*/
+using TupleType = Cabana::Tuple<DataTypes>;
+
+/*
+  Create a pointer of TupleType, which will be used in Fortran
+ */
+TupleType * particle = new TupleType;
+
+/*
+  Declare functions that will be mixed with Fortran
+ */
+extern "C" {
+  void tupleExample(TupleType *); //written in Fortan; called by C++
+  void delete_tuple();            //written in C++; called by Fortan
```

Comment:
Why is there not an equivalent `allocate_tuple` function that returns a pointer we can store in some Fortran data structure? That could get us away from this global pointer issue. How would one normally allocate, store, and then finally delete persistent data in an application with Fortran?

Reply:
My understanding with casting between C++ std::tuple and C struct (or Fortran derived type) is that tuple to struct is easy but the other way is not (but I could be totally wrong).  In Fortran, one usually use allocatable arrays, one calls allocate() and deallocate() for memory management. I guess some experiments are needed to see if what you suggested can be realized.

Reply:
Per our last conference call, we will figure out the answer in another pull request.